File: tmp/run_light_rag_eval.js
const fs = require("fs");
const path = require("path");

const ROOT = path.resolve(__dirname, "..");
const CORPUS_PATH = path.join(ROOT, "docs", "light_rag_corpus.json");
const DATASET_PATH = path.join(ROOT, "docs", "light_rag_eval_dataset.json");
const RESULT_JSON_PATH = path.join(ROOT, "tmp", "light_rag_eval_results.json");
const RESULT_MD_PATH = path.join(ROOT, "tmp", "light_rag_eval_results.md");

function loadJson(filePath) {
  return JSON.parse(fs.readFileSync(filePath, "utf8"));
}

function tokenize(text) {
  const lowered = String(text || "").toLowerCase();
  const wordTokens = lowered.match(/[a-z0-9\-_]+/g) || [];
  const chineseChars = [...lowered].filter((ch) => ch >= "\u4e00" && ch <= "\u9fff");
  return [...wordTokens, ...chineseChars];
}

function normalizeText(text) {
  return String(text || "").trim().toLowerCase().replace(/\s+/g, " ");
}

function scoreEntry(query, entry) {
  const queryNorm = normalizeText(query);
  const queryTokens = new Set(tokenize(query));
  const haystack = [
    entry.title || "",
    entry.summary || "",
    (entry.keywords || []).join(" "),
    entry.id || "",
  ].join(" ");
  const haystackTokens = new Set(tokenize(haystack));

  let overlap = 0;
  for (const token of queryTokens) {
    if (haystackTokens.has(token)) overlap += 1;
  }
  let score = overlap;

  for (const keyword of entry.keywords || []) {
    const keywordNorm = normalizeText(keyword);
    if (keywordNorm && queryNorm.includes(keywordNorm)) {
      score += 3.0;
    } else {
      const keywordTokens = new Set(tokenize(keyword));
      let localOverlap = 0;
      for (const token of queryTokens) {
        if (keywordTokens.has(token)) localOverlap += 1;
      }
      score += 0.25 * localOverlap;
    }
  }

  const titleNorm = normalizeText(entry.title || "");
  if (titleNorm && queryNorm.includes(titleNorm)) {
    score += 4.0;
  }

  const entryId = normalizeText(entry.id || "");
  const tail = entryId.split(".").pop();
  if (tail && queryNorm.includes(tail)) {
    score += 2.0;
  }

  if (entry.type === "rule_doc") {
    if (["why", "difference", "rule", "rules", "interface", "architecture", "bridge"].some((token) => queryNorm.includes(token))) {
      score += 0.2;
    }
  } else if (entry.type === "case") {
    if (["case", "example", "regression", "tc"].some((token) => queryNorm.includes(token))) {
      score += 0.2;
    }
  }

  return score;
}

function rankEntries(query, corpusEntries) {
  return corpusEntries
    .map((entry) => ({
      id: entry.id,
      type: entry.type,
      title: entry.title,
      score: Number(scoreEntry(query, entry).toFixed(4)),
    }))
    .sort((a, b) => (b.score - a.score) || a.id.localeCompare(b.id));
}

function reciprocalRank(rankedIds, expectedIds) {
  for (let index = 0; index < rankedIds.length; index += 1) {
    if (expectedIds.has(rankedIds[index])) {
      return 1 / (index + 1);
    }
  }
  return 0;
}

function precisionAtK(rankedIds, expectedIds, k) {
  if (k <= 0) return 0;
  const topIds = rankedIds.slice(0, k);
  if (topIds.length === 0) return 0;
  const hits = topIds.filter((entryId) => expectedIds.has(entryId)).length;
  return hits / topIds.length;
}

function recallAtK(rankedIds, expectedIds, k) {
  if (expectedIds.size === 0) return 1;
  const hits = rankedIds.slice(0, k).filter((entryId) => expectedIds.has(entryId)).length;
  return hits / expectedIds.size;
}

function summarize(metricLists) {
  const summary = {};
  for (const [name, values] of Object.entries(metricLists).sort(([a], [b]) => a.localeCompare(b))) {
    summary[name] = values.length ? Number((values.reduce((sum, value) => sum + value, 0) / values.length).toFixed(4)) : 0;
  }
  return summary;
}

function evaluate(corpus, dataset, kValues = [1, 3, 5]) {
  const entries = corpus.entries;
  const samples = dataset.samples;
  const overallMetrics = {};
  const categoryMetrics = {};
  const perSampleResults = [];

  function pushMetric(bucket, name, value) {
    if (!bucket[name]) bucket[name] = [];
    bucket[name].push(value);
  }

  for (const sample of samples) {
    const expectedIds = new Set([...(sample.expected_rule_ids || []), ...(sample.expected_case_ids || [])]);
    const ranked = rankEntries(sample.query, entries);
    const rankedIds = ranked.map((item) => item.id);
    const metrics = {};

    const rr = reciprocalRank(rankedIds, expectedIds);
    pushMetric(overallMetrics, "mrr", rr);
    if (!categoryMetrics[sample.category]) categoryMetrics[sample.category] = {};
    pushMetric(categoryMetrics[sample.category], "mrr", rr);
    metrics.mrr = Number(rr.toFixed(4));

    for (const k of kValues) {
      const hit = rankedIds.slice(0, k).some((entryId) => expectedIds.has(entryId)) ? 1 : 0;
      const precision = precisionAtK(rankedIds, expectedIds, k);
      const recall = recallAtK(rankedIds, expectedIds, k);

      pushMetric(overallMetrics, `hit_rate@${k}`, hit);
      pushMetric(overallMetrics, `precision@${k}`, precision);
      pushMetric(overallMetrics, `recall@${k}`, recall);

      pushMetric(categoryMetrics[sample.category], `hit_rate@${k}`, hit);
      pushMetric(categoryMetrics[sample.category], `precision@${k}`, precision);
      pushMetric(categoryMetrics[sample.category], `recall@${k}`, recall);

      metrics[`hit_rate@${k}`] = Number(hit.toFixed(4));
      metrics[`precision@${k}`] = Number(precision.toFixed(4));
      metrics[`recall@${k}`] = Number(recall.toFixed(4));
    }

    perSampleResults.push({
      id: sample.id,
      category: sample.category,
      query: sample.query,
      expected_ids: [...expectedIds].sort(),
      top_hits: ranked.slice(0, 5),
      metrics,
    });
  }

  const byCategory = {};
  for (const [category, metrics] of Object.entries(categoryMetrics).sort(([a], [b]) => a.localeCompare(b))) {
    byCategory[category] = summarize(metrics);
  }

  return {
    generated_at_utc: new Date().toISOString(),
    corpus_size: entries.length,
    sample_count: samples.length,
    k_values: kValues,
    overall: summarize(overallMetrics),
    by_category: byCategory,
    samples: perSampleResults,
  };
}

function renderMarkdown(result) {
  const lines = [];
  lines.push("# Lightweight RAG Eval Results", "");
  lines.push(`- Generated at (UTC): \`${result.generated_at_utc}\``);
  lines.push(`- Corpus size: \`${result.corpus_size}\``);
  lines.push(`- Sample count: \`${result.sample_count}\``, "");
  lines.push("## Overall", "");
  for (const [metric, value] of Object.entries(result.overall)) {
    lines.push(`- \`${metric}\`: \`${value}\``);
  }
  lines.push("", "## By Category", "");
  for (const [category, metrics] of Object.entries(result.by_category)) {
    lines.push(`### ${category}`);
    for (const [metric, value] of Object.entries(metrics)) {
      lines.push(`- \`${metric}\`: \`${value}\``);
    }
    lines.push("");
  }
  lines.push("## Sample Highlights", "");
  for (const sample of result.samples.slice(0, 8)) {
    lines.push(`### ${sample.id} - ${sample.category}`);
    lines.push(`- Query: ${sample.query}`);
    lines.push(`- Expected: ${sample.expected_ids.join(", ")}`);
    lines.push(`- Top 3: ${sample.top_hits.slice(0, 3).map((item) => `${item.id} (${item.score})`).join(", ")}`);
    lines.push(`- MRR: \`${sample.metrics.mrr}\``, "");
  }
  return lines.join("\n");
}

function main() {
  const corpus = loadJson(CORPUS_PATH);
  const dataset = loadJson(DATASET_PATH);
  const result = evaluate(corpus, dataset);

  fs.writeFileSync(RESULT_JSON_PATH, JSON.stringify(result, null, 2), "utf8");
  fs.writeFileSync(RESULT_MD_PATH, renderMarkdown(result), "utf8");

  console.log("Lightweight RAG evaluation finished.");
  console.log(`Corpus size: ${result.corpus_size}`);
  console.log(`Sample count: ${result.sample_count}`);
  for (const [metric, value] of Object.entries(result.overall)) {
    console.log(`${metric}: ${value}`);
  }
  console.log(`Saved JSON: ${RESULT_JSON_PATH}`);
  console.log(`Saved Markdown: ${RESULT_MD_PATH}`);
}

main();

File: tmp/run_light_rag_eval.py
from __future__ import annotations

import json
import re
from collections import defaultdict
from datetime import datetime, timezone
from pathlib import Path
from typing import Any


ROOT = Path(__file__).resolve().parents[1]
CORPUS_PATH = ROOT / "docs" / "light_rag_corpus.json"
DATASET_PATH = ROOT / "docs" / "light_rag_eval_dataset.json"
RESULT_JSON_PATH = ROOT / "tmp" / "light_rag_eval_results.json"
RESULT_MD_PATH = ROOT / "tmp" / "light_rag_eval_results.md"


def load_json(path: Path) -> dict[str, Any]:
    return json.loads(path.read_text(encoding="utf-8"))


def tokenize(text: str) -> list[str]:
    lowered = text.lower()
    word_tokens = re.findall(r"[a-z0-9\-\_]+", lowered)
    chinese_chars = [ch for ch in lowered if "\u4e00" <= ch <= "\u9fff"]
    return word_tokens + chinese_chars


def normalize_text(text: str) -> str:
    return re.sub(r"\s+", " ", text.strip().lower())


def score_entry(query: str, entry: dict[str, Any]) -> float:
    query_norm = normalize_text(query)
    query_tokens = set(tokenize(query))
    haystack_parts = [
        entry.get("title", ""),
        entry.get("summary", ""),
        " ".join(entry.get("keywords", []) or []),
        entry.get("id", ""),
    ]
    haystack = " ".join(haystack_parts)
    haystack_tokens = set(tokenize(haystack))

    overlap = len(query_tokens.intersection(haystack_tokens))
    score = float(overlap)

    for keyword in entry.get("keywords", []) or []:
        keyword_norm = normalize_text(str(keyword))
        if keyword_norm and keyword_norm in query_norm:
            score += 3.0
        else:
            keyword_tokens = set(tokenize(str(keyword)))
            if keyword_tokens:
                score += 0.25 * len(query_tokens.intersection(keyword_tokens))

    title_norm = normalize_text(entry.get("title", ""))
    if title_norm and title_norm in query_norm:
        score += 4.0

    entry_id = normalize_text(entry.get("id", ""))
    if entry_id and entry_id.split(".")[-1] in query_norm:
        score += 2.0

    if entry.get("type") == "rule_doc":
        if any(token in query_norm for token in ("why", "difference", "rule", "rules", "interface", "architecture", "bridge")):
            score += 0.2
    elif entry.get("type") == "case":
        if any(token in query_norm for token in ("case", "example", "regression", "tc")):
            score += 0.2

    return score


def rank_entries(query: str, corpus_entries: list[dict[str, Any]]) -> list[dict[str, Any]]:
    ranked = []
    for entry in corpus_entries:
        score = score_entry(query, entry)
        ranked.append(
            {
                "id": entry["id"],
                "type": entry["type"],
                "title": entry["title"],
                "score": round(score, 4),
            }
        )
    ranked.sort(key=lambda item: (-item["score"], item["id"]))
    return ranked


def reciprocal_rank(ranked_ids: list[str], expected_ids: set[str]) -> float:
    for index, entry_id in enumerate(ranked_ids, start=1):
        if entry_id in expected_ids:
            return 1.0 / index
    return 0.0


def precision_at_k(ranked_ids: list[str], expected_ids: set[str], k: int) -> float:
    if k <= 0:
        return 0.0
    top_ids = ranked_ids[:k]
    if not top_ids:
        return 0.0
    hits = sum(1 for entry_id in top_ids if entry_id in expected_ids)
    return hits / len(top_ids)


def recall_at_k(ranked_ids: list[str], expected_ids: set[str], k: int) -> float:
    if not expected_ids:
        return 1.0
    hits = sum(1 for entry_id in ranked_ids[:k] if entry_id in expected_ids)
    return hits / len(expected_ids)


def evaluate(corpus: dict[str, Any], dataset: dict[str, Any], k_values: tuple[int, ...] = (1, 3, 5)) -> dict[str, Any]:
    entries = corpus["entries"]
    samples = dataset["samples"]

    overall_metrics: dict[str, list[float]] = defaultdict(list)
    category_metrics: dict[str, dict[str, list[float]]] = defaultdict(lambda: defaultdict(list))
    per_sample_results: list[dict[str, Any]] = []

    for sample in samples:
        expected_ids = set(sample.get("expected_rule_ids", []) + sample.get("expected_case_ids", []))
        ranked = rank_entries(sample["query"], entries)
        ranked_ids = [item["id"] for item in ranked]

        sample_result = {
            "id": sample["id"],
            "category": sample["category"],
            "query": sample["query"],
            "expected_ids": sorted(expected_ids),
            "top_hits": ranked[:5],
            "metrics": {},
        }

        rr = reciprocal_rank(ranked_ids, expected_ids)
        overall_metrics["mrr"].append(rr)
        category_metrics[sample["category"]]["mrr"].append(rr)
        sample_result["metrics"]["mrr"] = round(rr, 4)

        for k in k_values:
            hit = 1.0 if any(entry_id in expected_ids for entry_id in ranked_ids[:k]) else 0.0
            precision = precision_at_k(ranked_ids, expected_ids, k)
            recall = recall_at_k(ranked_ids, expected_ids, k)

            overall_metrics[f"hit_rate@{k}"].append(hit)
            overall_metrics[f"precision@{k}"].append(precision)
            overall_metrics[f"recall@{k}"].append(recall)

            category_metrics[sample["category"]][f"hit_rate@{k}"].append(hit)
            category_metrics[sample["category"]][f"precision@{k}"].append(precision)
            category_metrics[sample["category"]][f"recall@{k}"].append(recall)

            sample_result["metrics"][f"hit_rate@{k}"] = round(hit, 4)
            sample_result["metrics"][f"precision@{k}"] = round(precision, 4)
            sample_result["metrics"][f"recall@{k}"] = round(recall, 4)

        per_sample_results.append(sample_result)

    def summarize(metric_lists: dict[str, list[float]]) -> dict[str, float]:
        return {
            name: round(sum(values) / len(values), 4) if values else 0.0
            for name, values in sorted(metric_lists.items())
        }

    category_summary = {
        category: summarize(metrics)
        for category, metrics in sorted(category_metrics.items())
    }

    return {
        "generated_at_utc": datetime.now(timezone.utc).isoformat(),
        "corpus_size": len(entries),
        "sample_count": len(samples),
        "k_values": list(k_values),
        "overall": summarize(overall_metrics),
        "by_category": category_summary,
        "samples": per_sample_results,
    }


def render_markdown(result: dict[str, Any]) -> str:
    lines: list[str] = []
    lines.append("# Lightweight RAG Eval Results")
    lines.append("")
    lines.append(f"- Generated at (UTC): `{result['generated_at_utc']}`")
    lines.append(f"- Corpus size: `{result['corpus_size']}`")
    lines.append(f"- Sample count: `{result['sample_count']}`")
    lines.append("")
    lines.append("## Overall")
    lines.append("")
    for metric, value in result["overall"].items():
        lines.append(f"- `{metric}`: `{value}`")
    lines.append("")
    lines.append("## By Category")
    lines.append("")
    for category, metrics in result["by_category"].items():
        lines.append(f"### {category}")
        for metric, value in metrics.items():
            lines.append(f"- `{metric}`: `{value}`")
        lines.append("")
    lines.append("## Sample Highlights")
    lines.append("")
    for sample in result["samples"][:8]:
        lines.append(f"### {sample['id']} - {sample['category']}")
        lines.append(f"- Query: {sample['query']}")
        lines.append(f"- Expected: {', '.join(sample['expected_ids'])}")
        lines.append(
            "- Top 3: "
            + ", ".join(f"{item['id']} ({item['score']})" for item in sample["top_hits"][:3])
        )
        lines.append(f"- MRR: `{sample['metrics']['mrr']}`")
        lines.append("")
    return "\n".join(lines)


def main() -> None:
    corpus = load_json(CORPUS_PATH)
    dataset = load_json(DATASET_PATH)
    result = evaluate(corpus, dataset)

    RESULT_JSON_PATH.write_text(json.dumps(result, ensure_ascii=False, indent=2), encoding="utf-8")
    RESULT_MD_PATH.write_text(render_markdown(result), encoding="utf-8")

    print("Lightweight RAG evaluation finished.")
    print(f"Corpus size: {result['corpus_size']}")
    print(f"Sample count: {result['sample_count']}")
    for metric, value in result["overall"].items():
        print(f"{metric}: {value}")
    print(f"Saved JSON: {RESULT_JSON_PATH}")
    print(f"Saved Markdown: {RESULT_MD_PATH}")


if __name__ == "__main__":
    main()
